### PR TITLE
Move Java50CleanUp to jdt.core.manipulation

### PR DESCRIPTION
**Pull request by datho7561**
## What it does
Move the functionality of `Java50CleanUp` and the classes it depends on to core.manipulations. This will allow for reusing it in eclipse.jdt.ls.

## How to test
Verify that you can access `Java50CleanUp` in a different project (eg. eclipse.jdt.ls) without needing to require any UI-related bundles.

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/Java50CleanUpCore.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
+/*******************************************************************************
```
Comment (datho7561) [inline org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/Java50CleanUpCore.java:1]: Looking at other similar changes, we didn't migrate the actual clean up to core manipulation. The reason is that most of the functionality this adds onto `Java50Fix` is stuff that's only helpful in the context of configuring the clean up using a UI. I'll put this class back into `eclipse.jdt.ui` when I get a chance.

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/JavaManipulationPlugin.java [hunk 2]
```diff
@@ -43,6 +43,8 @@ public class JavaManipulationPlugin extends Plugin implements DebugOptionsListen
 
 	public static boolean DEBUG_AST_PROVIDER;
 
+	public static boolean DEBUG_TYPE_CONSTRAINTS;
```
Comment (rgrunber) [inline org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/JavaManipulationPlugin.java:46]: Would it make sense to remove https://github.com/eclipse-jdt/eclipse.jdt.ui/blob/939c9911577144372fff1d15b8211c49012ff221/org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/JavaPlugin.java#L166 ? You've migrated all the points in the code that would have used that boolean.

File: org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/fix/Java50CleanUp.java [hunk 3]
```diff
@@ -81,11 +85,12 @@ protected ICleanUpFix createFix(CompilationUnit compilationUnit, IProblemLocatio
 
 		boolean addAnnotations= isEnabled(CleanUpConstants.ADD_MISSING_ANNOTATIONS);
 		boolean addOverride= isEnabled(CleanUpConstants.ADD_MISSING_ANNOTATIONS_OVERRIDE);
-		return Java50Fix.createCleanUp(compilationUnit, problems,
+		ICleanUpFixCore cleanUpFixCore= Java50FixCore.createCleanUp(compilationUnit, (IProblemLocationCore[]) problems,
```
Comment (rgrunber) [inline org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/fix/Java50CleanUp.java:88]: This kind of cast can't be done. To reproduce go into Java -> Compiler -> Errors/Warnings preferences, and set "Missing @Override annotation" as an Error. Next create some class that implements some interface where you omit the @Override annotations. Now, from the problems view, try to quick-fix 2 or more of the selected annotation errors. A dialog should appear and that should activate the method that applies it on the array of problem locations (as opposed to the whole compilation unit).

<details>
<summary>stacktrace</summary>

```
Root exception:
java.lang.ClassCastException: class [Lorg.eclipse.jdt.ui.text.java.IProblemLocation; cannot be cast to class [Lorg.eclipse.jdt.internal.ui.text.correction.IProblemLocationCore; ([Lorg.eclipse.jdt.ui.text.java.IProblemLocation; is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @10167267; [Lorg.eclipse.jdt.internal.ui.text.correction.IProblemLocationCore; is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @69318a32)
	at org.eclipse.jdt.internal.ui.fix.Java50CleanUp.createFix(Java50CleanUp.java:88)
	at org.eclipse.jdt.internal.ui.fix.AbstractMultiFix.createFix(AbstractMultiFix.java:47)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring.calculateChange(CleanUpRefactoring.java:772)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpASTRequestor.calculateSolutions(CleanUpRefactoring.java:301)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpASTRequestor.acceptAST(CleanUpRefactoring.java:279)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.resolve(CompilationUnitResolver.java:1008)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.resolve(CompilationUnitResolver.java:692)
	at org.eclipse.jdt.core.dom.ASTParser.createASTs(ASTParser.java:954)
	at org.eclipse.jdt.internal.corext.dom.ASTBatchParser.createASTs(ASTBatchParser.java:82)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpFixpointIterator.next(CleanUpRefactoring.java:399)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring.cleanUpProject(CleanUpRefactoring.java:682)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring.checkFinalConditions(CleanUpRefactoring.java:642)
	at org.eclipse.ltk.core.refactoring.Refactoring.checkAllConditions(Refactoring.java:165)
	at org.eclipse.jdt.internal.ui.refactoring.RefactoringExecutionHelper$Operation.run(RefactoringExecutionHelper.java:84)
	at org.eclipse.jdt.internal.core.BatchOperation.executeOperation(BatchOperation.java:41)
	at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:740)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2380)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2405)
	at org.eclipse.jdt.core.JavaCore.run(JavaCore.java:5955)
	at org.eclipse.jdt.internal.ui.actions.WorkbenchRunnableAdapter.run(WorkbenchRunnableAdapter.java:108)
	at org.eclipse.jdt.internal.ui.text.correction.proposals.FixCorrectionProposal.lambda$0(FixCorrectionProposal.java:206)
	at org.eclipse.jdt.internal.ui.refactoring.RefactoringExecutionHelper.perform(RefactoringExecutionHelper.java:187)
	at org.eclipse.jdt.internal.ui.refactoring.RefactoringExecutionHelper.perform(RefactoringExecutionHelper.java:151)
	at org.eclipse.jdt.internal.ui.text.correction.proposals.FixCorrectionProposal.resolve(FixCorrectionProposal.java:211)
	at org.eclipse.jdt.internal.ui.text.correction.CorrectionMarkerResolutionGenerator$CorrectionMarkerResolution.run(CorrectionMarkerResolutionGenerator.java:152)
	at org.eclipse.ui.internal.views.markers.QuickFixWizard.processResolution(QuickFixWizard.java:123)
	at org.eclipse.ui.internal.views.markers.QuickFixWizard.lambda$0(QuickFixWizard.java:98)
	at org.eclipse.jface.operation.ModalContext.runInCurrentThread(ModalContext.java:434)
	at org.eclipse.jface.operation.ModalContext.run(ModalContext.java:352)
	at org.eclipse.jface.wizard.WizardDialog.run(WizardDialog.java:1034)
	at org.eclipse.ui.internal.views.markers.QuickFixWizard.performFinish(QuickFixWizard.java:100)
	at org.eclipse.jface.wizard.WizardDialog.finishPressed(WizardDialog.java:833)
	at org.eclipse.jface.wizard.WizardDialog.buttonPressed(WizardDialog.java:473)
	at org.eclipse.jface.dialogs.Dialog.lambda$0(Dialog.java:619)
	at org.eclipse.swt.events.SelectionListener$1.widgetSelected(SelectionListener.java:84)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:252)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5815)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1529)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:5025)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4477)
	at org.eclipse.jface.window.Window.runEventLoop(Window.java:823)
	at org.eclipse.jface.window.Window.open(Window.java:799)
	at org.eclipse.ui.internal.views.markers.QuickFixHandler.execute(QuickFixHandler.java:160)
	at org.eclipse.ui.internal.handlers.HandlerProxy.execute(HandlerProxy.java:283)
	at org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(E4HandlerProxy.java:97)
	at jdk.internal.reflect.GeneratedMethodAccessor263.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:58)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:317)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:251)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:173)
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:156)
	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:488)
	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:485)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:213)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.executeCommand(KeyBindingDispatcher.java:308)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.press(KeyBindingDispatcher.java:580)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.processKeyEvent(KeyBindingDispatcher.java:647)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.filterKeySequenceBindings(KeyBindingDispatcher.java:439)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher$KeyDownFilter.handleEvent(KeyBindingDispatcher.java:96)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.filterEvent(Display.java:1908)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1528)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1555)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1538)
	at org.eclipse.swt.widgets.Widget.sendKeyEvent(Widget.java:1577)
	at org.eclipse.swt.widgets.Widget.gtk_key_press_event(Widget.java:937)
	at org.eclipse.swt.widgets.Control.gtk_key_press_event(Control.java:4013)
	at org.eclipse.swt.widgets.Composite.gtk_key_press_event(Composite.java:921)
	at org.eclipse.swt.widgets.Tree.gtk_key_press_event(Tree.java:2372)
	at org.eclipse.swt.widgets.Widget.windowProc(Widget.java:2434)
	at org.eclipse.swt.widgets.Control.windowProc(Control.java:6834)
	at org.eclipse.swt.widgets.Tree.windowProc(Tree.java:4285)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:6123)
	at org.eclipse.swt.internal.gtk3.GTK3.gtk_main_do_event(Native Method)
	at org.eclipse.swt.widgets.Display.eventProc(Display.java:1552)
	at org.eclipse.swt.internal.gtk3.GTK3.gtk_main_iteration_do(Native Method)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4474)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:643)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:550)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:171)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)
```
</details>

File: org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/fix/Java50CleanUp.java [hunk 4]
```diff
@@ -67,25 +70,33 @@ addAnotations && isEnabled(CleanUpConstants.ADD_MISSING_ANNOTATIONS_DEPRECATED)
...
 	@Override
 	protected ICleanUpFix createFix(CompilationUnit compilationUnit, IProblemLocation[] problems) throws CoreException {
 		if (compilationUnit == null)
 			return null;
 
+		ProblemLocationCore[] coreProblems= new ProblemLocationCore[problems.length];
+		for (int i= 0; i < coreProblems.length; i++) {
+			coreProblems[i]= new ProblemLocationCore(problems[i].getOffset(), problems[i].getLength(), problems[i].getProblemId(), problems[i].getProblemArguments(), problems[i].isError(),
+					problems[i].getMarkerType());
+		}
```
Comment (rgrunber) [inline org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/fix/Java50CleanUp.java:90]: Also at  https://github.com/eclipse-jdt/eclipse.jdt.ui/blob/344ae32efe1d334d37635899ec2455a6aa665685/org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/fix/CleanUpFixWrapper.java#L56-L59 . More common to go from IProblem -> IProblemLocationCore .

Is it worth adding an extra constructor to ProblemLocationCore that takes an IProblemLocation ?

File: org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest1d5.java [hunk 5]
```diff
@@ -4246,4 +4268,63 @@ public void testConvertStringBufferToStringBuilderAll() throws Exception {
...
+		ASTParser parser= ASTParser.newParser(IASTSharedValues.SHARED_AST_LEVEL);
+		parser.setResolveBindings(true);
+		parser.setProject(getProject());
+		final CompilationUnit[] roots= new CompilationUnit[1];
+		parser.createASTs(new ICompilationUnit[] { cu1 }, new String[0], new ASTRequestor() {
+			@Override
+			public void acceptAST(ICompilationUnit source, CompilationUnit ast) {
+				roots[0]= ast;
+			}
+		}, null);
```
Comment (rgrunber) [inline org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest1d5.java:4323]: I think you can use `ASTNode createAST(IProgressMonitor monitor)` from `ASTParser`. Although it returns an `ASTNode`, you can check for yourself that it's always the `CompilationUnit` (ast root).

Comment (rgrunber) [reply org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest1d5.java:4323]: Oh and you can use `parser.setSource(..)` to provide the `ICompilationUnit`

Comment (datho7561) [reply org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest1d5.java]: Thanks!

**Comment by rgrunber** [comment]
Also, can you rename the commit to :

```
Move Java50Fix to jdt.core.manipulation

- Add test case for activating cleanup as a multi quick-fix
```

File: org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest1d5.java [hunk 6]
```diff
@@ -4246,4 +4267,73 @@ public void testConvertStringBufferToStringBuilderAll() throws Exception {
 				new HashSet<>(Arrays.asList(MultiFixMessages.StringBufferToStringBuilderCleanUp_description)));
 	}
 
+	@Test
+	public void testOverrideMultiFix() throws Exception {
+		Hashtable<String, String> opts= JavaCore.getOptions();
+		String oldOverrideError = opts.get(JavaCore.COMPILER_PB_MISSING_OVERRIDE_ANNOTATION);
```
Comment (rgrunber) [inline org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest1d5.java:4273]: Is this necessary ? I know you want to reset to the old value so as to not disturb other tests, but according to https://github.com/eclipse-jdt/eclipse.jdt.ui/blob/344ae32efe1d334d37635899ec2455a6aa665685/org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTestCase.java#L95 that should already happen. This class does inherit from there.

Comment (datho7561) [reply org.eclipse.jdt.ui.tests/ui/org/eclipse/jdt/ui/tests/quickfix/CleanUpTest1d5.java:4273]: Good catch!

File: org.eclipse.jdt.core.manipulation/META-INF/MANIFEST.MF [hunk 7]
```diff
@@ -49,7 +50,11 @@ Export-Package: org.eclipse.jdt.core.manipulation,
  org.eclipse.jdt.internal.corext.refactoring.structure;x-friends:="org.eclipse.jdt.ui",
  org.eclipse.jdt.internal.corext.refactoring.surround;x-friends:="org.eclipse.jdt.ui",
  org.eclipse.jdt.internal.corext.refactoring.tagging;x-friends:="org.eclipse.jdt.ui",
- org.eclipse.jdt.internal.corext.refactoring.typeconstraints.types;manipulation=split;mandatory:=manipulation;x-friends:="org.eclipse.jdt.ui",
+ org.eclipse.jdt.internal.corext.refactoring.generics;x-friends:="org.eclipse.jdt.ui",
```
Comment (rgrunber) [inline org.eclipse.jdt.core.manipulation/META-INF/MANIFEST.MF:53]: Ugh, just noticed this. Missing the `manipulation=split;mandatory:=manipulation`. I would preserve that.

Comment (datho7561) [reply org.eclipse.jdt.core.manipulation/META-INF/MANIFEST.MF:53]: shoot don't merge yet it I did it wrong

Comment (datho7561) [reply org.eclipse.jdt.core.manipulation/META-INF/MANIFEST.MF]: okay. Should be fixed

**Pull request merged**